Add hl-indent-scope

ideasman42 opened the pull request:
### Brief summary of what the package does

Highlight indentation, unlike existing packages it uses scope depth from the syntax-table (works well for C/C++ / GLSL Lisp and similar languages), CMake & Python are also supported.

Using the syntax-table has the advantage that pre-processor commands don't break the highlighting, something that made existing packages such such as [highlight-indent-guides](https://github.com/DarthFennec/highlight-indent-guides) unusable for me.
Screenshot of `hl-indent-scope` https://codeberg.org/attachments/b50b1aeb-b653-4ca0-9a27-2fcc311d9bab

### Direct link to the package repository

https://codeberg.org/ideasman42/emacs-hl-indent-scope

### Your association with the package

Maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm with `x`: -->

- [X] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [X] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [X] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [X] My elisp byte-compiles cleanly
- [X] `M-x checkdoc` is happy with my docstrings
- [X] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [ ] I have confirmed some of these without doing them

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  This looks cool, but I'm obligated to ask whether the pre-processor issue (which I'm assuming only affects C/C++? but correct me if I'm wrong) couldn't be treated as a bug that could be raised to the `highlight-indent-guides` maintainers?  It would obviously be a great outcome for everyone if this issue could be fixed there.

Also, when you say this works well for C/C++/Lisp and similar languages, does this *not* work well for e.g. whitespace-indented languages like e.g. Python or yaml? I ask so that I can get a better look at the context surrounding these -- thanks for your patience with the ignorant questions. :)

In the meantime, quick first pass --

`hl-indent-scope.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- hl-indent-scope.el#L393: It's safer to sharp-quote function names; use `#'`
- hl-indent-scope.el#L301: Consider `beginning-of-line`
- hl-indent-scope.el#L307: Consider `beginning-of-line`
- hl-indent-scope.el#L339: Consider `beginning-of-line`
```

ideasman42 commented:
Corrected errors _(really need to integrate melpazoid into my checking utility)_

Yes, I can see why it might be good to extend `highlight-indent-guides` or similar packages, in the case of `pre-processor`, most cases would be solved by checking for lines that match a pre-processor expression, although a fool-proof check is more involved as it requires checking that line isn't a continuation of a macro on the previous line or part of a comment.

In my opinion this package works differently enough that it's not exactly a fix on other packages, more a different package that has pros/cons, but in some cases is far more useful - in short, whenever there is a chance some lines don't follow indentation - these don't impact indentation display.

There is a more general difference - that is - _"brackets define indentation display"_ instead of _"indentation display is read from indentation"_.

For C/C++/Java/Rust ... etc, only showing indentation from blocks enclosed by curly braces has the advantage of differentiating multi-line blocks from wrapped function calls or assignments.

Personally I find this introduces less visual noise and it's quicker to visually scan down a document, although the difference is admittedly not huge.

Using the syntax-tree has some potential for different kinds of features, it's relatively trivial (for e.g. that indentation highlighting depends on the kind of block - so wrapped function calls could have a different color to C++ lambda's / class definition / enum / struct ... etc). That's something I considered but wanted to first get the basics working.

----

Regarding Python/YAML, right - it doesn't work for these languages currently (it works for nested dictionaries, lists, tuples .. but not functions/classes), I'd like to support them but that would involve extracting a tree structure from text, that tree structure could then be used in-place of tree typically extracted from Emacs syntax-table.

ideasman42 commented:
Update, lisp like languages are now working well:

This image shows the difference between highlighting indentation & scope.
![screen](https://user-images.githubusercontent.com/1869379/182262397-f4dd8950-8554-4cc2-a035-3ee0bace91e1.png)

riscy commented:
Thanks for the explanation and the helpful screenshots.

Since this is restricted in the languages can be applied, it may be worth mentioning that somewhere in the `;;; Commentary` that the focus is languages with C/C++ or lisp-like syntax.  I don't intend to apply too fine a comb, but while this looks beautiful in some rust code I quickly looked at, it feels like there are maybe a few kinks to work out for the lisp languages:

<img width="504" alt="image" src="https://user-images.githubusercontent.com/1696710/183307208-257073c0-b718-4b33-956a-d5a1634d572f.png">
<img width="521" alt="image" src="https://user-images.githubusercontent.com/1696710/183307214-eeb04fec-e2ef-4bbc-b79a-43678836f751.png">

Ruby code with mixed curly braces and do/end also presents some artifacts:

<img width="279" alt="image" src="https://user-images.githubusercontent.com/1696710/183307347-87e86e83-bb77-48ad-a645-ffb0bf37c9c4.png">

Of course these weren't languages that you were targeting, but I think we should be careful here: it can be tempting to index a package with a speculative/general name before it actually has general applicability, and I think that's not good for users.  And since this package is also "close" to what a few existing packages do (up to their lack of handling for preprocessor directives?) it really needs to stand out as novel.

On that note, I think the name `hl-indent-scope` is a bit of a misnomer.  This is not highlighting where the *indentation* goes (and maybe this your differentiator vs highlight-indent-guides).  Instead, this is highlighting *scope* inasmuch as it's defined by supported delimiters like curly braces and parentheses.  `hl-scope` makes more sense, but something like `hl-c-scope` (as a very crude example) leaves no ambiguity about where this package can actually be applied.

Hopefully that wasn't too verbose, and my concerns are clear.

riscy commented:
I forgot to include that each of the files should have license boilerplate.  Currently:

- unknown license -- hl-indent-scope-preset-c++-mode.el: C++ preset
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to hl-indent-scope-preset-c++-mode.el
- unknown license -- hl-indent-scope-preset-c-mode.el: C preset
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to hl-indent-scope-preset-c-mode.el
- unknown license -- hl-indent-scope-preset-cmake-mode.el: CMake preset
- Add *formal* license boilerplate or an [SPDX-License-Identifier](https://spdx.dev/ids/) to hl-indent-scope-preset-cmake-mode.el
- GPL-3.0-or-later -- hl-indent-scope.el: Highlight indentation by scope

ideasman42 commented:
Thanks for more detailed review, added SPDX licenses and updated the `;;; Commentary`.

The issues you've pointed out in Lisp are caused by tabs, lisp code without tabs should work perfectly well (screen-shot after running `untabify`).

![screen](https://user-images.githubusercontent.com/1869379/183317690-3c4d4df5-f577-47ae-93cd-90d835fa8298.png)

Looking into this and tabs could be supported, it's just currently they aren't, I tend to avoid tabs as they aren't, other similar packages also don't support tabs https://melpa.org/#/highlight-indentation - but I can see this could be a show stopper for some developers.

As for naming the package based on a language, this is awkward, this does in fact highlight indentation, it's just that the indentation level is delimited by nested scope.

As for your point about languages, as I do intend to support Python (next on my TODO list for this package), and there isn't a reason other languages couldn't be supported (although it depends on the complexity of parsing their syntax).

If C-like, Lisp-like & Python (which is one of the most popular languages that doesn't use braces) are supported - naming this a particular language is also misleading :/ - off hand this leaves out Ruby/Pascal/Shell/Haskell/XML/HTML/Makefiles (+ others that might not be so well known).

Nevertheless your points are well taken, and I'm not so keen on including this package based on what it may support in the future.

Perhaps supporting Python + tab support could be completed for this package to be considered general enough not to require it being named based on a single language/syntax.




----

ideasman42 commented:
Update: Python is now supported too.

riscy commented:
Thanks very much for the explanations - I think this has lots of potential to diverge from the other indentation-highlighting modes (different colored blocks especially sounds interesting).  I'll leave these two very minor lints while I merge:

`hl-indent-scope-preset-python-mode.el` with checkdoc 0.6.2:
```
hl-indent-scope-preset-python-mode.el:249: Argument `flat-keywords' should appear (as FLAT-KEYWORDS) in the doc string
```

- hl-indent-scope-preset-python-mode.el#L312: Consider `beginning-of-line`

ideasman42 commented:
Thanks, resolved the lint (renaming variables missed doc-strings).
